### ISSUE TITLE
Feature request: native Windows ML inference

### ISSUE THREAD

**Issue by ESchouten** [ultralytics/ultralytics#23517]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

I request a native Windows ML inference backend (device='winml').

Currently, hardware-accelerated inference on Windows (AMD/Intel/NPU) requires shipping the heavy onnxruntime-directml and manually configuring execution providers.

This proposed backend would leverage the OS-native Windows.AI.MachineLearning API. By creating sessions with LearningModelDeviceKind.Default, Ultralytics can delegate hardware selection entirely to Windows. This allows the OS to automatically route inference to the best available device—whether it is a Discrete GPU, a specific NPU (Copilot+ PCs), or the CPU—without requiring the library to bundle runtimes or manage complex hardware detection logic.

### Use case

**Zero-Dependency Deployment:** Developers can build lightweight Windows desktop applications by using the inference engine already built into the OS, eliminating the need to bundle large onnxruntime or torch binaries.
**Automatic NPU & GPU Support:** Users on modern hardware (e.g., Surface Pro 11, AMD Radeon) get immediate hardware acceleration. The OS dynamically manages the "Hybrid Loop," creating a path for Ultralytics to run efficiently on the 1+ billion Windows devices.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ESchouten, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for the detailed feature request around a native Windows inference backend (`device='winml'`) leveraging Windows.AI.MachineLearning 🙌 This is an automated response, and an Ultralytics engineer will also assist soon to review feasibility, scope, and potential integration paths.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the request—today Ultralytics routes inference backends primarily by exported model format via `AutoBackend`, and there isn’t a Windows.AI.MachineLearning (WinML) execution path yet (see the [AutoBackend reference](https://docs.ultralytics.com/reference/nn/autobackend/)). A workable path forward would be a Windows-only, optional `winml` backend branch in `ultralytics/nn/autobackend.py` that loads an exported ONNX model and returns outputs with the same shapes/order as our existing ONNX/ORT path, so postprocessing stays unchanged; if you can share a minimal Python snippet (and the specific WinRT/Python binding you’d expect us to depend on) showing how you create a `LearningModelSession` and read/write tensors efficiently, we can evaluate feasibility and API design in this issue.

**Comment by ESchouten** [comment]
What I end up with;
```python
import asyncio
import numpy as np
import winrt.windows.ai.machinelearning as winml
import winrt.windows.storage as ws

# PyWinRT deps:
#   pip install winrt-windows-ai-machinelearning winrt-windows-storage

async def load_session(model_path: str):
    file = await ws.StorageFile.get_file_from_path_async(model_path)
    model = await winml.LearningModel.load_from_storage_file_async(file)
    device = winml.LearningModelDevice(winml.LearningModelDeviceKind.DEFAULT)
    return model, winml.LearningModelSession(model, device)

def run(session, model, x: np.ndarray):
    x = np.ascontiguousarray(x, dtype=np.float32)
    tensor = winml.TensorFloat.create_from_iterable([int(d) for d in x.shape], x.ravel().tolist())

    binding = winml.LearningModelBinding(session)
    binding.bind(model.input_features[0].name, tensor)

    results = session.evaluate(binding, "run_id")

    # Preserve output order to match ONNX/ORT
    outputs = []
    for f in model.output_features:
        out = results.outputs[f.name]
        data = np.array(out.get_as_vector_view(), dtype=np.float32)
        outputs.append(data.reshape(list(out.shape)))
    return outputs

async def main():
    model, session = await load_session(r"C:\models\yolo.onnx")
    tensor = np.zeros((1, 3, 640, 640), dtype=np.float32)
    outputs = run(session, model, dummy)
    print([o.shape for o in outputs])
```

**Comment by glenn-jocher** [comment]
Thanks for sharing this—this is a solid starting point, but for it to be viable inside Ultralytics we’ll need to avoid the `x.ravel().tolist()` path (it creates ~1.2M Python floats for `1x3x640x640`, which will dominate runtime/memory), so can you confirm whether PyWinRT exposes any buffer/array-based tensor constructor that can ingest a contiguous `np.float32` blob without converting to a Python list? Also note your example has a small typo (`dummy` is undefined; you likely meant `tensor`). If you can post a tiny benchmark + feature-name/order dump using an Ultralytics-exported ONNX (and the exact `pip` deps/versions) in a self-contained snippet per our [Minimum Reproducible Example guide](https://docs.ultralytics.com/help/minimum-reproducible-example/), we can better judge API/packaging and how to map WinML outputs to our existing ONNX output order.

**Comment by ESchouten** [comment]
The following snippet uses a WinRT buffer (TensorFloat.create_from_buffer) to avoid creating the ~1.2M Python floats.
At the moment I sadly don't have a Windows machine at hand to benchmark the snippet.

```python
import asyncio
import numpy as np
import winrt.windows.ai.machinelearning as winml
import winrt.windows.storage as ws
import winrt.windows.storage.streams as streams

# PyWinRT deps:
#   pip install winrt-windows-ai-machinelearning winrt-windows-storage

async def load_session(model_path: str):
    file = await ws.StorageFile.get_file_from_path_async(model_path)
    model = await winml.LearningModel.load_from_storage_file_async(file)
    device = winml.LearningModelDevice(winml.LearningModelDeviceKind.DEFAULT)
    return model, winml.LearningModelSession(model, device)

def numpy_to_tensor_float(x: np.ndarray) -> winml.TensorFloat:
    x = np.ascontiguousarray(x, dtype=np.float32)
    shape = [int(d) for d in x.shape]

    data = x.tobytes()
    buffer = streams.Buffer(len(data))
    writer = streams.DataWriter()
    writer.write_bytes(data)
    buffer = writer.detach_buffer()

    return winml.TensorFloat.create_from_buffer(shape, buffer)

def run(session, model, x: np.ndarray):
    tensor = numpy_to_tensor_float(x)

    binding = winml.LearningModelBinding(session)
    binding.bind(model.input_features[0].name, tensor)

    results = session.evaluate(binding, "run_id")

    outputs = []
    for f in model.output_features:
        out = results.outputs[f.name]
        data = np.array(out.get_as_vector_view(), dtype=np.float32)
        outputs.append(data.reshape(list(out.shape)))
    return outputs

async def main():
    model, session = await load_session(r"C:\models\yolo.onnx")
    tensor = np.zeros((1, 3, 640, 640), dtype=np.float32)
    outputs = run(session, model, tensor)
    print([o.shape for o in outputs])
```

**Comment by glenn-jocher** [comment]
`TensorFloat.create_from_buffer()` is the right direction, but we still need to confirm whether the `DataWriter.write_bytes(x.tobytes())` path is doing an extra full copy (and whether PyWinRT exposes any way to wrap a contiguous `np.ndarray`/`memoryview` as an `IBuffer` without reallocating), and likewise whether outputs can be read back as a raw buffer instead of `get_as_vector_view()` to avoid another large copy; if you (or anyone) can run a quick timing + feature-name dump against a Ultralytics-exported ONNX, we can evaluate what a clean `winml` branch would look like in the existing [AutoBackend source](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/autobackend.py).  

```python
print("inputs:", [f.name for f in model.input_features])
print("outputs:", [f.name for f in model.output_features])
```

**Comment by ESchouten** [comment]
@glenn-jocher Seems like there is an "onnxruntime-windowsml" since last month, so perhaps we don't have to implement it ourselves.

https://pypi.org/project/onnxruntime-windowsml/

**Comment by glenn-jocher** [comment]
Good find—[onnxruntime-windowsml (PyPI)](https://pypi.org/project/onnxruntime-windowsml/) exposes the standard `onnxruntime` Python API backed by Windows ML, so we may be able to support “native Windows ML” by using our existing ONNX path but creating the `InferenceSession` *without* forcing a `providers=[...]` list (so Windows ML can pick the best EP/device). ([pypi.org](https://pypi.org/project/onnxruntime-windowsml/))

If you (or anyone) can test on Windows 11 and paste `ort.__version__`, `ort.get_available_providers()`, plus a quick `session.run()` timing (and confirm any required Windows App SDK init noted in the Windows ML docs), that would tell us whether a small `device='winml'` switch is sufficient or if a dedicated backend is still needed. ([learn.microsoft.com](https://learn.microsoft.com/en-us/windows/ai/new-windows-ml/api-reference))

```bash
pip uninstall -y onnxruntime onnxruntime-gpu
pip install onnxruntime-windowsml
```

```python
import time
import numpy as np
import onnxruntime as ort

sess = ort.InferenceSession("yolo.onnx")  # intentionally no providers=...
print("ort:", ort.__version__)
print("providers:", ort.get_available_providers())

x = np.zeros((1, 3, 640, 640), dtype=np.float32)
t0 = time.time()
sess.run(None, {sess.get_inputs()[0].name: x})
print("ms:", (time.time() - t0) * 1e3)
```

**Comment by ESchouten** [comment]
@glenn-jocher 
The following is with a half, np.float16, model
```
ort: 1.24.0
providers: ['DmlExecutionProvider', 'CPUExecutionProvider']
ms: 404.766321182251
```

**Comment by glenn-jocher** [comment]
Thanks—`404 ms` for a single `640x640` run looks like either first-run compilation overhead or partial CPU fallback; could you paste the output of `sess.get_providers()` (actual providers in-use, not just available), `sess.get_inputs()[0].type`, and an average over multiple warmed-up runs per the [ONNX Runtime Python API docs](https://onnxruntime.ai/docs/api/python/api_summary.html)?

```python
import time
import numpy as np
import onnxruntime as ort

sess = ort.InferenceSession("yolo.onnx")  # keep default provider selection
print("ort:", ort.__version__)
print("available:", ort.get_available_providers())
print("in-use:", sess.get_providers())
print("input:", sess.get_inputs()[0].name, sess.get_inputs()[0].type)

x = np.zeros((1, 3, 640, 640), dtype=np.float16)  # match fp16 model input type if required
sess.run(None, {sess.get_inputs()[0].name: x})  # warmup

t0 = time.perf_counter()
for _ in range(20):
    sess.run(None, {sess.get_inputs()[0].name: x})
ms = (time.perf_counter() - t0) * 1e3 / 20
print("avg ms:", ms)
```

**Comment by ESchouten** [comment]
@glenn-jocher Your code snippet ran on the CPU

I ran the following snippet:
```python
import time
import numpy as np
import onnxruntime as ort

print("ort:", ort.__version__)
print("available:", ort.get_available_providers())

model_path = "yolo.onnx"
input_name = None

# Force DML; remove CPU fallback to ensure you don't silently run on CPU.
providers = ["DmlExecutionProvider"]  # no fallback

sess = ort.InferenceSession(model_path, providers=providers)

print("in-use:", sess.get_providers())
print("provider options:", sess.get_provider_options())

inp = sess.get_inputs()[0]
input_name = inp.name
print("input:", input_name, inp.type, inp.shape)

# Keep dtype consistent with the model's input type
x = np.zeros((1, 3, 640, 640), dtype=np.float16)

# warmup
sess.run(None, {input_name: x})

t0 = time.perf_counter()
for _ in range(20):
    sess.run(None, {input_name: x})
ms = (time.perf_counter() - t0) * 1e3 / 20
print("avg ms:", ms)
```

With the following result:
```
ort: 1.24.0
available: ['DmlExecutionProvider', 'CPUExecutionProvider']
2026-02-10 21:51:30.0490547 [W:onnxruntime:, session_state.cc:1316 onnxruntime::VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.
2026-02-10 21:51:30.0582330 [W:onnxruntime:, session_state.cc:1318 onnxruntime::VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
in-use: ['DmlExecutionProvider', 'CPUExecutionProvider']
provider options: {'DmlExecutionProvider': {}, 'CPUExecutionProvider': {}}
input: images tensor(float16) ['batch', 3, 'height', 'width']
avg ms: 13.387814996531233
```

**Comment by ESchouten** [comment]
@glenn-jocher btw, instead of running on DirectML, on Windows 11, windows ml can also download execution providers so it can run on e.g. nvidia tensorrt, openvino or amd with the same executable.

https://learn.microsoft.com/en-us/windows/ai/new-windows-ml/get-started?tabs=python#step-2-download-and-register-eps

https://learn.microsoft.com/en-us/windows/ai/new-windows-ml/supported-execution-providers?source=recommendations

**Comment by glenn-jocher** [comment]
That `~13.4 ms` warmed-up result with `DmlExecutionProvider` in `sess.get_providers()` suggests the ORT+Windows ML route is viable; it’s also normal for `CPUExecutionProvider` to still appear because ORT will intentionally keep some ops on CPU even when DML is the primary EP. If you register the downloadable Windows ML EPs (TensorRT/OpenVINO/etc), can you rerun `ort.get_available_providers()` and `sess.get_providers()` *without* forcing `providers=[...]` and paste the output here—if those EPs show up and are selected automatically, we can likely implement `device='winml'` as an ONNX-only path that avoids hardcoding providers and also updates our dependency check to accept `onnxruntime-windowsml` on Windows (today we check for the `onnxruntime` wheel name). For the current “explicit backend” approach, the supported route remains exporting to dedicated formats (e.g. OpenVINO/TensorRT) as summarized in the [Model deployment options guide](https://docs.ultralytics.com/guides/model-deployment-options/).

**Comment by ESchouten** [comment]
Yes, that is what I do in my monkeypatched ultralytics application
```python
import onnxruntime as ort

def InferenceSession(path_or_bytes, sess_options=None, providers=None, **kwargs):
        providers = ort.get_available_providers()
        LOGGER.info("ORT providers available: %s", providers)
        return _InferenceSession(path_or_bytes, sess_options, providers, **kwargs)

ort.InferenceSession = InferenceSession
```
DmlExecutionProvider comes up here before CPUExecutionProvider so that works great. With onnxruntime-gpu TensorrtExecutionProvider comes up before CUDA which comes before the rest so that works as well.

**Comment by glenn-jocher** [comment]
Got it—your monkeypatch confirms that “use whatever ORT reports as available, in that priority order” is a workable policy on Windows; we can implement this cleanly in `ultralytics/nn/autobackend.py` behind an ONNX-only `device='winml'` switch (i.e., set `providers=ort.get_available_providers()` instead of our current CPU/CUDA forcing) and update our dependency check to treat the `onnxruntime-windowsml` wheel as satisfying the ONNX backend on Windows. When you get a chance, please paste `yolo checks` from that same environment (so we can verify the exact Ultralytics import path + versions) and confirm whether registering the downloadable Windows ML EPs changes `ort.get_available_providers()` order/contents as expected; that’ll tell us how “automatic” we can make it without extra knobs beyond the standard ONNX flow described in the [Export mode docs](https://docs.ultralytics.com/modes/export/).
